examples/todoapp/lib/base/data_sources/local/todo_local_data_source.dart

```diff
@@ -0,0 +1,167 @@
+    realm.write(() {
+      realm.deleteMany(todos as List<TodoModel>);
+    });
+  }
+
+  $TodoModel addTodo($TodoModel todo) {
+    realm.write(() {
+      realm.add<TodoModel>(todo as TodoModel);
+    });
+    return todo;
+  }
+
+  List<TodoModel> fetchAllUnsyncedTodos() {
+    final results = realm.query<TodoModel>(r'synced == $0', [false]);
+    return results.map((element) => element).toList();
+  }
+
+  List<TodoModel> fetchAllTodos() {
+    final results = realm.query<TodoModel>(r'action != $0', ['delete']);
+    return results.map((element) => element).toList();
+  }
+
+  $TodoModel updateTodoById(String id, $TodoModel todo) {
+    final result = realm.find<TodoModel>(id);
+
+    if (result == null) {
+      throw Exception('Todo with id $id not found');
+    }
+    realm.write(() {
+      result.title = todo.title;
+      result.description = todo.description;
+      result.completed = todo.completed;
+      result.synced = todo.synced;
+      result.action = todo.action;
+    });
+    return result;
+  }
+
+  TodoModel updateCompletedById(
+    String id,
+    bool completed, {
+    bool synced = true,
+    String? action,
+  }) {
+    final todo = realm.find<TodoModel>(id);
+
+    if (todo == null) {
+      throw Exception('Todo with id $id not found');
+    }
+    realm.write(() {
+      todo.completed = completed;
+      todo.synced = synced;
+      todo.action = action ?? TodoModelActions.none.name;
+    });
+    return todo;
+  }
+
+  List<TodoModel> updateCompletedForAll(
+    bool completed, {
+    bool synced = true,
+    String? action,
+  }) {
+    final results = realm.all<TodoModel>().query('completed != $completed');
+    final todos = results.toList();
+    realm.write(() {
+      for (var todo in todos) {
+        todo.completed = completed;
+        todo.action = action ?? TodoModelActions.none.name;
+        todo.synced = synced;
+      }
+    });
+    return todos.map((element) => element).toList();
+  }
+
+  List<TodoModel> softDeleteCompleted({bool synced = true, String? action}) {
+    final results = realm.all<TodoModel>().query('completed == true');
+    final todos = results.toList();
+    realm.write(() {
+      for (var todo in todos) {
+        ///Instead of deleting the todo, we will mark it for deletion
+        todo.action = action ?? TodoModelActions.none.name;
+        todo.synced = synced;
+      }
+    });
+    return todos;
+  }
+
+  List<TodoModel> deleteCompleted({bool synced = true, String? action}) {
+    final results = realm.all<TodoModel>().query('completed == true');
+    final todos = results.toList();
+    realm.write(() {
+      for (var todo in todos) {
+        realm.delete<TodoModel>(todo);
+      }
+    });
+    return todos;
+  }
+
+  void deleteTodoById(
+    String id, {
+    bool synced = true,
+    String? action,
+  }) {
+    final todo = realm.find<TodoModel>(id);
+
+    if (todo == null) {
+      throw Exception('Todo with id $id not found');
```

Comment:
Same as above. Please use the NotFoundErrorModel

Reply:
Done